streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -82,6 +82,68 @@ abstract class ZStream[-R, +E, +O](
   ): ZStream[R1, E1, O1] =
     ZStream.absolve(ev(self))
 
+  /**
+   * Aggregates elements of this stream using the provided sink for as long
+   * as the downstream operators on the stream are busy.
+   *
+   * This operator divides the stream into two asynchronous "islands". Operators upstream
+   * of this operator run on one fiber, while downstream operators run on another. Whenever
+   * the downstream fiber is busy processing elements, the upstream fiber will feed elements
+   * into the sink until it signals completion.
+   *
+   * Any sink can be used here, but see [[Sink.foldWeightedM]] and [[Sink.foldUntilM]] for
+   * sinks that cover the common usecases.
+   */
+  final def aggregateAsync[R1 <: R, E1 >: E, O1 >: O, P](transducer: ZTransducer[R1, E1, O1, P]): ZStream[R1, E1, P] =
+    ZStream {
+      for {
+        pull   <- self.process
+        push   <- transducer.push
+        bucket <- ZQueue.bounded[Take[E1, P]](1).toManaged(_.shutdown)
+        done   <- ZRef.makeManaged(false)
+        produce = {
+          def finish: ZIO[R1, Nothing, Boolean] =
+            push(None)
+              .repeat(Schedule.doWhile(_.isEmpty))
+              .foldCauseM(
+                Cause
+                  .sequenceCauseOption(_)
+                  .fold(bucket.offer(Take.End))(c => bucket.offer(Exit.halt(c.map(Some(_))))),
+                ps => bucket.offer(Exit.succeed(ps))
+              )
+              .as(false)
+
+          def maybeDrain(os: Chunk[O]): ZIO[R1, Nothing, Boolean] =
+            push(Some(os))
+              .foldCauseM(
+                Cause.sequenceCauseOption(_).fold(IO.unit)(c => bucket.offer(Exit.halt(c.map(Some(_)))).unit),
+                ps => bucket.offer(Exit.succeed(ps))
```

Comment:
Another observation - this implementation allows the operator to get started on processing the next chunk of data before the previous chunk was taken from the queue. Arguable whether this is desirable or not, but it’s different from the previous implementation.